### PR TITLE
bpo-21536: C extensions are no longer linked to libpython

### PR DESCRIPTION
**Pull request by vstinner**
On Unix, C extensions are no longer linked to libpython.

It is now possible to load a C extension built using a shared library
Python with a statically linked Python.

When Python is embedded, libpython must not be loaded with
RTLD_LOCAL, but RTDL_GLOBAL instead. Previously, using RTLD_LOCAL, it
was already not possible to load C extensions which were not linked
to libpython, like C extensions of the standard library built by the
"*shared*" section of Modules/Setup.

distutils, python-config and python-config.py have been modified.

https://bugs.python.org/issue21536

### CODE REVIEW

**Comment by vstinner** [comment]
This PR allows to load a C extension built by a shared libpython with a statically linked Python: https://bugs.python.org/issue21536#msg340819

It also allows to load C extension built in release mode with a Python built in debug mode: https://bugs.python.org/issue21536#msg340821

**Comment by vstinner** [comment]
One potential issue: linking an application to libpython to embed Python, see discussion at https://bugs.python.org/issue21536#msg340837

**Review by pitrou — APPROVED** [review]
LGTM. Thank you!

File: Doc/whatsnew/3.8.rst [hunk 1]
```diff
@@ -845,16 +845,19 @@ Changes in the Python API
 Changes in the C API
 --------------------
 
+* On Unix, C extensions are no longer linked to libpython. When Python is
+  embedded, ``libpython`` must not be loaded with ``RTLD_LOCAL``, but
+  ``RTDL_GLOBAL`` instead. Previously, using ``RTLD_LOCAL``, it was already not
```
Comment (skoslowski) [inline Doc/whatsnew/3.8.rst]: minor typo ``RTLD_GLOBAL``

Comment (vstinner) [reply Doc/whatsnew/3.8.rst]: Oops, fixed :-)

**Comment by froody** [comment]
This change causes strange bugs on Mac OSX due to the two-level namespace symbol resolution for dynamic libraries. A C extension that links against libpython will resolve symbols from libpython instead of the python executable (for example, https://github.com/PDAL/python/issues/75 ). I believe linking C extensions on Mac OS X with "-flat_namespace" would be a reliable fix, as it would allow the python executable symbols to take precedence over those from libpython.

**Comment by vstinner** [comment]
Please don't comment closed PRs, but open a new issue at https://bugs.python.org/ Otherwise, your comment will simply be ignored.

**Pull request merged**